File: DIP/dip_intact.cpp
#include <iostream>
using namespace std;

// Abstraction (Interface)
class Database {
public:
    virtual void save(string data) = 0;
};

// MySQL implementation (Low-level module)
class MySQLDatabase : public Database {
public:
    // Save data to SQL
    void save(string data) override {
        cout << "Executing SQL Query: INSERT INTO users VALUES('" << data << "');" << endl;
    }
};

// MongoDB implementation (Low-level module)
class MongoDBDatabase : public Database {
public:
    // Save data to MongoDB
    void save(string data) override {
        cout << "Executing MongoDB Function: db.users.insert({name: '" << data << "'})" << endl;
    }
};

// High-level module, it is now loosely coupled
class UserService {
private:
    Database* db;  

public:
        UserService(Database* database) {  
        db = database;
    }
    
    // Store user in the database
    void store_user(string user) {
        db->save(user);
    }
};

int main() {
    MySQLDatabase mysql;
    MongoDBDatabase mongodb;

    UserService service1(&mysql);
    service1.store_user("User 1");

    UserService service2(&mongodb);
    service2.store_user("User 2");
}

File: DIP/java/DIPIntact.java
// Abstraction
interface Database {
    void save(String data);
}

// MySQL implementation
class MySQLDatabase implements Database {
    public void save(String data) {
        System.out.println("Executing SQL Query: INSERT INTO users VALUES('" + data + "');");
    }
}

// MongoDB implementation
class MongoDBDatabase implements Database {
    public void save(String data) {
        System.out.println("Executing MongoDB Function: db.users.insert({name: '" + data + "'})");
    }
}

// High-level module loosely coupled
class UserService {
    private Database db;

    public UserService(Database db) {
        this.db = db;
    }

    public void storeUser(String user) {
        db.save(user);
    }
}

public class DIPIntact {
    public static void main(String[] args) {
        MySQLDatabase mysql = new MySQLDatabase();
        MongoDBDatabase mongodb = new MongoDBDatabase();

        UserService service1 = new UserService(mysql);
        service1.storeUser("User 1");

        UserService service2 = new UserService(mongodb);
        service2.storeUser("User 2");
    }
}
